### PR TITLE
Add smoke test for binary package

### PR DESCRIPTION
**Pull request by guyang3532**

### CODE REVIEW

File: windows/internal/smoke_test.bat [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    exit /b 1
+)
+call "%VS15VCVARSALL%" x64
+
+set install_root=%CD%
+set INCLUDE=%INCLUDE%;%install_root%/include;%install_root%/include/torch/csrc/api/include
+set LIB=%LIB%;%install_root%/lib
+set PATH=%PATH%;%install_root%/lib
+
+>example-app.cpp (
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:135]: nit: Consider creating a new file?

Comment (peterjc123) [reply windows/internal/smoke_test.bat:135]: And the name of the executable is not clear.

File: windows/internal/smoke_test.bat [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+echo    return 0;
+echo }
+)
+
+del example-app.exe
+cl example-app.cpp c10.lib torch_cpu.lib /EHsc
+.\example-app.exe
+if ERRORLEVEL 1 exit /b 1
+
+echo Checking that MKL is available
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:148]: And here

File: windows/internal/smoke_test.bat [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+if ERRORLEVEL 1 exit /b 1
+
+if "%NVIDIA_GPU_EXISTS%" == "0" (
+    echo "Skip CUDA tests for machines without a Nvidia GPU card"
+    goto end
+)
+
+del example-app.cpp
+>example-app.cpp (
+echo #include ^<torch/torch.h^>
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:170]: And here

File: windows/internal/smoke_test.bat [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+curl -k https://repo.anaconda.com/miniconda/Miniconda3-latest-Windows-x86_64.exe -o "%miniconda_exe%"
+call ..\conda\install_conda.bat
+if ERRORLEVEL 1 exit /b 1
+
+set "PATH=%CONDA_HOME%;%CONDA_HOME%\scripts;%CONDA_HOME%\Library\bin;%PATH%"
+
+conda create -qyn testenv python=%DESIRED_PYTHON%
+source activate testenv >/dev/null
+for /F "delims=" %%i in ('where /R "%PYTORCH_FINAL_PACKAGE_DIR:/=\%" *.tar.bz2') do conda install -y "%%i" --offline
+conda install -yq future numpy protobuf six
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:68]: nit: Check the exitcode here?

File: windows/internal/smoke_test.bat [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+set /a CUDA_VER=%CUDA_VERSION%
+set CUDA_VER_MAJOR=%CUDA_VERSION:~0,-1%
+set CUDA_VER_MINOR=%CUDA_VERSION:~-1,1%
+set CUDA_VERSION_STR=%CUDA_VER_MAJOR%.%CUDA_VER_MINOR%
+conda install -yq -c pytorch "cudatoolkit=%CUDA_VERSION_STR%"
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:74]: nit: Check the exitcode here?

File: windows/internal/smoke_test.bat [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+goto end
+
+:libtorch
+echo "install and test libtorch"
+
+powershell internal/vs_install.ps1
+if ERRORLEVEL 1 exit /b 1
+
+for /F "delims=" %%i in ('where /R "%PYTORCH_FINAL_PACKAGE_DIR:/=\%" *-latest.zip') do unzip "%%i" -d tmp
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:112]: nit: Consider using 7z instead. `unzip` is only available in Bash. The equivalent line using 7z is `7z x "%%i" -otmp`.

File: windows/internal/smoke_test.bat [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+curl --retry 3 -kL "%PYTHON_INSTALLER_URL%" --output python-amd64.exe
+if errorlevel 1 exit /b 1
+
+start /wait "" python-amd64.exe /quiet InstallAllUsers=1 PrependPath=1 Include_test=0 TargetDir=%CD%\Python%PYTHON_VERSION%
+if errorlevel 1 exit /b 1
+
+set "PATH=%CD%\Python%PYTHON_VERSION%\Scripts;%CD%\Python%PYTHON_VERSION%;%PATH%"
+
+for /F "delims=" %%i in ('where /R "%PYTORCH_FINAL_PACKAGE_DIR:/=\%" *.whl') do pip install "%%i"
+pip install -q future numpy protobuf six "mkl>=2019"
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:46]: nit: Check exitcode here？

File: windows/internal/smoke_test.bat [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+rmdir /s /q conda
+del miniconda.exe
+curl -k https://repo.anaconda.com/miniconda/Miniconda3-latest-Windows-x86_64.exe -o "%miniconda_exe%"
+call ..\conda\install_conda.bat
+if ERRORLEVEL 1 exit /b 1
+
+set "PATH=%CONDA_HOME%;%CONDA_HOME%\scripts;%CONDA_HOME%\Library\bin;%PATH%"
+
+conda create -qyn testenv python=%DESIRED_PYTHON%
+source activate testenv >/dev/null
```
Comment (peterjc123) [inline windows/internal/smoke_test.bat:66]: Again this is `Bash` only, please use `call %CONDA_HOME%\condabin\activate.bat testenv > NUL`.

File: windows/internal/update_driver.bat [hunk 9]
```diff
@@ -0,0 +1,4 @@
+set "DRIVER_DOWNLOAD_LINK=https://s3.amazonaws.com/ossci-windows/442.50-tesla-desktop-winserver-2019-2016-international.exe"
+curl -k -L DRIVER_DOWNLOAD_LINK --output 442.50-tesla-desktop-winserver-2019-2016-international.exe
```
Comment (peterjc123) [inline windows/internal/update_driver.bat:2]: nit:
1. check error code
2. add `--retry` option

File: windows/internal/update_driver.bat [hunk 10]
```diff
@@ -0,0 +1,4 @@
...
+
+start /wait 442.50-tesla-desktop-winserver-2019-2016-international.exe -s -noreboot
```
Comment (peterjc123) [inline windows/internal/update_driver.bat:4]: nit: check exit code

File: windows/internal/update_driver.bat (hunk 9 above)
Comment (peterjc123) [inline windows/internal/update_driver.bat:2]:
```suggestion
curl --retry 3 -kL "%DRIVER_DOWNLOAD_LINK%" --output 442.50-tesla-desktop-winserver-2019-2016-international.exe
if errorlevel 1 exit /b 1
```

File: check_binary.sh [hunk 11]
```diff
@@ -256,8 +256,8 @@ build_and_run_example_cpp () {
   if [ -f "${install_root}/lib/libtorch_cuda.so" ] || [ -f "${install_root}/lib/libtorch_cuda.dylib" ]; then
     TORCH_CUDA_LINK_FLAGS="-ltorch_cuda"
   fi
-  g++ example-app.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o example-app
-  ./example-app
+  g++ ${BUILDER_ROOT}/test_example_code/$1.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o $1
```
Comment (peterjc123) [inline check_binary.sh:259]: Why is `$1` used here?

File: windows/internal/driver_update.bat [hunk 12]
```diff
@@ -0,0 +1,9 @@
+set "DRIVER_DOWNLOAD_LINK=https://s3.amazonaws.com/ossci-windows/442.50-tesla-desktop-winserver-2019-2016-international.exe"
+curl --retry 3 -kL %DRIVER_DOWNLOAD_LINK% --output 442.50-tesla-desktop-winserver-2019-2016-international.exe
+if errorlevel 1 exit /b 1
+
+start /wait 442.50-tesla-desktop-winserver-2019-2016-international.exe -s -noreboot
+if errorlevel 1 exit /b 1
+
+del 442.50-tesla-desktop-winserver-2019-2016-international.exe || ver > NUL
+
```
Comment (peterjc123) [inline windows/internal/driver_update.bat:9]: nit: empty line?

**Comment by peterjc123** [comment]
Could you please take some time to review this PR, @seemethere?

File: check_binary.sh (hunk 11 above)
Comment (guyang3532) [reply check_binary.sh]: The code file name is passed as argument from outside.

Comment (guyang3532) [reply check_binary.sh]: > Why is `$1` used here?
The code file name is passed as argument from outside.

File: check_binary.sh [hunk 13]
```diff
@@ -256,8 +256,8 @@ build_and_run_example_cpp () {
   if [ -f "${install_root}/lib/libtorch_cuda.so" ] || [ -f "${install_root}/lib/libtorch_cuda.dylib" ]; then
     TORCH_CUDA_LINK_FLAGS="-ltorch_cuda"
   fi
-  g++ example-app.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o example-app
-  ./example-app
+  g++ /builder/test_example_code/$1.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o $1
```
Comment (peterjc123) [inline check_binary.sh:259]: Wow, just noticed that this is an absolute path.

File: check_binary.sh [hunk 14]
```diff
@@ -287,7 +287,7 @@ build_example_cpp_with_incorrect_abi () {
   if [ -f "${install_root}/lib/libtorch_cuda.so" ] || [ -f "${install_root}/lib/libtorch_cuda.dylib" ]; then
     TORCH_CUDA_LINK_FLAGS="-ltorch_cuda"
   fi
-  g++ example-app.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o example-app
+  g++ /builder/test_example_code/$1.cpp -I${install_root}/include -I${install_root}/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=$GLIBCXX_USE_CXX11_ABI -std=gnu++14 -L${install_root}/lib ${REF_LIB} ${ADDITIONAL_LINKER_FLAGS} -ltorch $TORCH_CPU_LINK_FLAGS $TORCH_CUDA_LINK_FLAGS $C10_LINK_FLAGS -o $1
```
Comment (peterjc123) [inline check_binary.sh:290]: and this

**Pull request merged**